lib/path/default.nix

```diff
@@ -94,6 +95,48 @@ let
 
 in /* No rec! Add dependencies on this file at the top. */ {
 
+  /* Append a subpath string to a path.
+
+    Like `path + ("/" + string)` but safer, because it errors instead of returning potentially surprising results.
+    More specifically, it checks that the first argument is a path value type (https://nixos.org/manual/nix/stable/language/values.html#type-path"),
+    and that the second argument is a valid subpath string (see `lib.path.subpath.isValid`).
+
+    Type:
+      append :: Path -> String -> Path
+
+    Example:
+      append /foo "bar/baz"
+      => /foo/bar/baz
+
+      # subpaths don't need to be normalised
+      append /foo "./bar//baz/./"
+      => /foo/bar/baz
+
+      # can append to root directory
+      append /. "foo/bar"
+      => /foo/bar
+
+      # first argument needs to be a path value type
+      append "/foo" "bar"
+      => <error>
+
+      # second argument needs to be a valid subpath string
+      append /foo /bar
+      => <error>
+      append /foo ""
+      => <error>
+      append /foo "/bar"
+      => <error>
+      append /foo "../bar"
+      => <error>
+  */
+  append = base: subpath:
+    assert assertMsg (isPath base) ''
+      lib.path.append: The first argument is of type ${builtins.typeOf base}, but a path was expected'';
```

Comment:
I'm putting `''` at the end of the string here and not on a newline because we don't want to print an extra newline (`throw` adds one). Apparently it doesn't matter though, because `throw` removes all trailing newlines anyways:

```
nix-repl> throw "hello\n\n\n\n\n"
error: hello

nix-repl>
```

I don't want to rely on that because it's undocumented and perhaps a bug. While I'm at it, there's a bug where `throw ""` doesn't print the red `error:` prefix:

```
nix-repl> throw ""                


nix-repl>
```